File: codes/python/polymorphism/subtype.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class Person(object):
    def talk(self):
        print("I am a person")

class Student(Person): pass

class Professor(Person): pass

class Robot(object):
    def talk(self):
        print("I am a robot")

import sys
def announce(msg, talker):
    sys.stdout.write(msg + ": ")
    talk = getattr(talker, "talk", None)
    if talk != None and callable(talk):
        talk()
    else:
        print("Sorry! I do not talk!")

pe = Person()
st = Student()
pr = Professor()
ro = Robot()

announce("Person", pe)
announce("Student", st)
announce("Professor", pr)
announce("Robot", ro)
announce("a number", 25)

print ("A table with different types")
elements = [pe, st, pr, ro, 25]
for i, element in enumerate(elements):
    try:
        element.talk()
    except Exception:
    	print("The element n°" + str(i) + " does not talk()")

File: codes/ruby/polymorphism/subtype.rb
#!/usr/bin/ruby -w

class Person
    def talk
        puts "I am a person"
    end
end

class Student < Person; end

class Professor < Person; end

class Robot
    def talk
        puts "I am a robot"
    end
end

def announce (msg, talker)
    print "#{msg}: "
    if talker.class.method_defined? :talk then
        talker.talk
    else
        puts "Sorry! I do not talk!"
    end
end

pe = Person.new
st = Student.new
pr = Professor.new
ro = Robot.new

announce("Person", pe)
announce("Student", st)
announce("Professor", pr)
announce("Robot", ro)
announce("a number", 25)

puts  "A table with different types"
elements = [pe, st, pr, ro, 25]
elements.each_with_index do |element, i|
    begin
        element.talk
    rescue Exception => e
        puts "The element n°#{i} does not talk()"
    end
end
